multiversx_sdk_core/transaction_factories/smart_contract_factory.py

```diff
@@ -0,0 +1,192 @@
+from pathlib import Path
+from typing import Any, List, Optional, Protocol, Tuple
+
+from Cryptodome.Hash import keccak
+
+from multiversx_sdk_core.address import Address
+from multiversx_sdk_core.constants import (ARGS_SEPARATOR,
+                                           CONTRACT_DEPLOY_ADDRESS,
+                                           DEFAULT_HRP,
+                                           TRANSACTION_OPTIONS_DEFAULT,
+                                           TRANSACTION_VERSION_DEFAULT,
+                                           VM_TYPE_WASM_VM)
+from multiversx_sdk_core.interfaces import (IAddress, IChainID, IGasLimit,
+                                            IGasPrice, INonce,
+                                            ITransactionValue)
+from multiversx_sdk_core.serializer import arg_to_string
+from multiversx_sdk_core.transaction import Transaction
+from multiversx_sdk_core.transaction_payload import TransactionPayload
+from multiversx_sdk_core.utils import read_binary_file
+
+
+class IConfig(Protocol):
+    chain_id: IChainID
+
+
+class CodeMetadata:
+    def __init__(self, upgradeable: bool = True, readable: bool = True, payable: bool = False, payable_by_sc: bool = False) -> None:
+        self.upgradeable = upgradeable
+        self.readable = readable
+        self.payable = payable
+        self.payable_by_sc = payable_by_sc
+
+    def to_hex(self) -> str:
+        flag_value_pairs = [
+            (0x01_00, self.upgradeable),
+            (0x04_00, self.readable),
+            (0x00_02, self.payable),
+            (0x00_04, self.payable_by_sc)
+        ]
+        metadata_value = self.sum_flag_values(flag_value_pairs)
+        return f"{metadata_value:04X}"
+
+    def sum_flag_values(self, flag_value_pairs: List[Tuple[int, bool]]) -> int:
+        value_sum = 0
+        for value, flag in flag_value_pairs:
+            if flag:
+                value_sum += value
+        return value_sum
+
+
+class SmartContractFactory:
+    def __init__(self, config: IConfig) -> None:
+        self.config = config
+
+    def deploy(self,
+               deployer: IAddress,
+               nonce: INonce,
+               bytecode_path: Path,
+               gas_limit: IGasLimit,
+               arguments: Optional[List[Any]] = None,
```

Comment:
Can default to empty list, without being optional? Also, below.

Reply:
Done.